### ISSUE TITLE
built-in support for the `file` protocol - i.e. must not rely on the `git` binary.

### ISSUE THREAD

**Issue by Byron** [Byron/gitoxide#734]

**Comment by NobodyXu** [comment]
Looking forward to this, it would be great if `gix` can support `--local` to use hardlinks to avoid copying `.git` or  `--shared` to avoid copying any objects at all when dealing with local repositories.

**Comment by Byron** [comment]
I agree, and it's so low-hanging that I might even do that just to warm up with the topic.

**Comment by weihanglo** [comment]
Sorry I didn't quite follow the recent updates in Gitoxide. Do we still shell out to `git-upload-pack` today?

**Comment by Byron** [comment]
Yes, that's the case. No work was done in that direction.

**Comment by weihanglo** [comment]
Yeah thanks for the info. I asked because I was wondering in Cargo if we can stabilize the shallow-clone support with the caveat of no native `file` protocol support (or just fallback to non-shallow)

**Comment by Byron** [comment]
> Yeah thanks for the info. I asked because I was wondering in Cargo if we can stabilize the shallow-clone support with the caveat of no native `file` protocol support (or just fallback to non-shallow)

I see - in the fallback case `file://` would probably  be served by `git2`, and then shallow isn't possible. Probably that's alright as shallow will be most useful when the URL isn't local.

**Comment by Byron** [comment]
Moved to [this Feature Proposal Discussion](https://github.com/GitoxideLabs/gitoxide/discussions/2771). This issue is being closed while the proposal is reviewed; it will become an issue again once review is complete.

_AI disclosure: This comment was written and posted by OpenAI Codex on behalf of Byron._
